cpp/src/join/quadtree_poly_filtering.cu

```diff
@@ -0,0 +1,292 @@
+  auto const node_offsets = quadtree.column(4);  // uint32_t
+
+  auto num_polys = poly_bbox.num_rows();
+
+  // Count the number of top-level nodes to start.
+  // This could be provided explicitly, but count_if should be fast enough.
+  auto num_top_level_children = thrust::count_if(rmm::exec_policy(stream)->on(stream),
+                                                 node_levels.begin<uint8_t>(),
+                                                 node_levels.end<uint8_t>(),
+                                                 thrust::placeholders::_1 == 0);
+
+  auto num_pairs = num_top_level_children * num_polys;
+
+  // The found poly-quad pairs are dynamic and can not be pre-allocated.
+  // Relevant arrays are resized accordingly for memory efficiency.
+
+  // Vectors for intermediate poly and node indices at each level
+  rmm::device_uvector<uint8_t> tmp_types(num_pairs, stream);
+  rmm::device_uvector<uint8_t> tmp_levels(num_pairs, stream);
+  rmm::device_uvector<uint32_t> tmp_node_idxs(num_pairs, stream);
+  rmm::device_uvector<uint32_t> tmp_poly_idxs(num_pairs, stream);
+
+  // Vectors for found pairs of poly and leaf node indices
+  rmm::device_uvector<uint8_t> out_types(num_pairs, stream);
+  rmm::device_uvector<uint8_t> out_levels(num_pairs, stream);
+  rmm::device_uvector<uint32_t> out_node_idxs(num_pairs, stream);
+  rmm::device_uvector<uint32_t> out_poly_idxs(num_pairs, stream);
+
+  // A zip iterator for the intermediate intersections or parent quadrants found during traversal
+  auto tmp_pairs = make_zip_iterator(
+    tmp_types.begin(), tmp_levels.begin(), tmp_node_idxs.begin(), tmp_poly_idxs.begin());
+
+  // A zip iterator for the intersecting poly and quad indicies found
+  auto out_pairs = make_zip_iterator(
+    out_types.begin(), out_levels.begin(), out_node_idxs.begin(), out_poly_idxs.begin());
+
+  cudf::size_type num_results{0};
+  cudf::size_type num_parents{0};
+  cudf::size_type num_children{0};
+
+  // Traverse the quadtree starting at level 0 and descending to `max_depth` or until no more parent
+  // quadrants are found.
+  for (cudf::size_type level{0}; level < max_depth; ++level) {
+    // Resize output device vectors and update the corresponding pointers. The next level will
+    // expand out to no more than `num_parents * 4` pairs, since a parent quadrant can have no more
+    // than 4 children.
+    size_t max_num_results = num_results + num_parents * 4;
+    if (max_num_results > out_types.capacity()) {
+      // grow preallocated output sizes in multiples of the current capacity
+      auto new_size = out_types.capacity() * ((max_num_results / out_types.capacity()) + 1);
+      out_types.resize(new_size, stream);
+      out_levels.resize(new_size, stream);
+      out_node_idxs.resize(new_size, stream);
+      out_poly_idxs.resize(new_size, stream);
+      out_pairs = make_zip_iterator(
+        out_types.begin(), out_levels.begin(), out_node_idxs.begin(), out_poly_idxs.begin());
+    }
```

Comment:
This is not done on level == 0. So can you factor level 0 out of the loop, and then have a simpler loop?

Reply:
Restructured the loop, hopefully it's a bit simpler.